### DISCUSSION TITLE
Can Nelua's preprocessor/macro be used alone (as a Lua library)?

### DISCUSSION THREAD

**Discussion by MuhammedZakir** [edubart/nelua-lang#189] [Q&A] [answered]
Mainly, I would like to use macros in configs. E.g. for NeoVim. Currently, I use Fennel for that, but I want to replace that with nelua.

NeoVim uses Lua 5.1 (because of LuaJIT). Would that create any problem?

if it helps, the workflow will be something like: `<name>.nelua` which will contain Lua code with Nelua macros -> compile to Lua using Nelua and save it to `<name>.lua` -> load that file whenever `<name>.nelua` is `require`d.

**Comment by edubart** [comment, marked as answer]
It is possible with lot of limitations and quirks, so I don't recommend that, thus I would say no.

Nelua can compile back to Lua, this allow using its preprocessor and type checking for Lua, however the Lua backend is not maintained anymore, it was done in the early days for parsing and testing purposes. Nowadays the Lua backend is unmaintained and lots of stuff will give compile time errors.

Here is an example of you asked:

```lua
## MYCONFIG = true
## if MYCONFIG then
  print 'hello from lua'
## end
```

```shell
$ nelua test.nelua --generator lua -o test.lua
$ cat test.lua 
-- Generated by Nelua 0.2.0-dev
-- Compile Hash: mFdHCahpdZ8woTtsei8ecRHJehC

print("hello from lua")
$ lua test.lua
hello from lua
```

So using `--generator` will make the compiler generate Lua code instead of C code. But as soon you start to make something more interesting, things start to fail:

```lua
## MYCONFIG = true
## if MYCONFIG then
  print 'hello world'
  local t = {}
  print(#t)
## end
```

```
$ nelua test.nelua --generator lua -o test.lua
^~~~~~~~~~~~~~~~~~
test.nelua:5:10: error: in unary operation `len`: invalid operation for type 'table'
  print(#t)
  
```

This happens because `#` is not defined for tables in the Lua backend, and the type analysis fails there. And the way Nelua preprocessor is made it is impossible to decouple it from the type checker analysis, so this requires many improvements in the type analyzer to handle Lua cases. However there is little to no interesting in meta programming or type checking Lua with the Nelua compiler, so I don't maintain that backend, Nelua is made mainly for compiling to C. This could be revisited in the future, but it is unlikely.

For you use case, remember that Lua is a dynamic language, you can meta program your configs in Lua itself, no need to plug in an extra compiling step to personalize Lua configurations.
